src/future/mod.rs

```diff
@@ -308,7 +308,40 @@ pub trait Future {
     fn boxed(self) -> BoxFuture<Self::Item, Self::Error>
         where Self: Sized + Send + 'static
     {
-        ::std::boxed::Box::new(self)
+        #[cfg(rust_at_least_1_18)]
+        fn boxed<F>(future: F) -> BoxFuture<F::Item, F::Error>
+            where F: Future + Sized + Send + 'static
+        {
+            use std::any::TypeId;
+            use std::mem;
+            use std::ptr;
+            use std::boxed;
+
+            // Prevent double boxing
+            if TypeId::of::<F>() == TypeId::of::<BoxFuture<F::Item, F::Error>>() {
+                assert!(mem::size_of::<F>() == mem::size_of::<BoxFuture<F::Item, F::Error>>());
+                unsafe {
+                    let mut r: BoxFuture<F::Item, F::Error> = mem::uninitialized();
+                    ptr::copy_nonoverlapping(
+                        &future as *const F,
+                        &mut r as *mut BoxFuture<F::Item, F::Error> as *mut u8 as *mut F,
+                        1);
+                    mem::forget(future);
+                    r
```

Comment:
No, because compiler doesn't know that `future` can be cast to `BoxFuture<...>`. Moreover, even `transmute` doesn't work, because compiler cannot prove that type sizes are equal.